README.md
# Proposed Pandoc::Elements extended selector expression syntax

This repository is the home for a **proposed** extended elements selector syntax for Pandoc::Elements
([on GitHub][PEonGH] and [on MetaCPAN][PEonMCPAN]).

**Please note that neither this syntax proposal nor the associated code has been adopted or endorsed by Pandoc::Elements or anybody else!**

In this repository you will find

*   [A description of the syntax proposal][description] as a Markdown file.

*   [Perl code which can parse selectors using this syntax][parser]
    as well as compile checker subroutines.

    You can test the syntax by running this script and typing selectors to STDIN.
    The selector will be compiled and the compiled subroutine or any error message
    will be printed to STDOUT.

Related discussion can be found
[here](https://github.com/jgm/pandoc/issues/4541)
and [here](https://github.com/nichtich/Pandoc-Elements/issues/18).

Feel free to open an [issue in this repository][issue] if you have any suggestion for an improvement *of this proposal*.  At the moment it is not likely that any new features will be added, but the wording and organization of the description can certainly be improved!  Please remember that the description must be valid GitHub Flavored Markdown *and* valid Pandoc Markdown!


[PEonGH]: https://github.com/nichtich/Pandoc-Elements
[PEonMCPAN]: https://metacpan.org/release/Pandoc-Elements
[description]: pandoc-elements-selectors-description.md
[parser]: pandoc-elements-selectors-parser.pl
[issue]: https://github.com/bpj/pandoc-elements-selectors/issues






CHANGES
CHANGELOG

<https://github.com/bpj/pandoc-elements-selectors>

2018-05-02 16:03:41 +0200
    speak of properties rather than methods (fixes #1)

        Benct Philip Jonsson <bpjonsson@gmail.com>

2018-04-25 17:33:25 +0200
    Document "shortcuts"/syntactic sugar more explicitly

        Benct Philip Jonsson <bpjonsson@gmail.com>

2018-04-25 16:08:12 +0200
    Document unquoted strings aka "words"

        Benct Philip Jonsson <bpjonsson@gmail.com>

2018-04-25 15:51:09 +0200
    change names in code to agree with terms in description

        Benct Philip Jonsson <bpjonsson@gmail.com>

2018-04-25 15:35:24 +0200
    Improve number regex

        Benct Philip Jonsson <bpjonsson@gmail.com>

2018-04-25 15:31:24 +0200
    Describe and motivate design principles and metacharacter choices

        Benct Philip Jonsson <bpjonsson@gmail.com>

2018-04-25 13:52:31 +0200
    Document the difference between subselectors and selector
    expressions better

        Benct Philip Jonsson <bpjonsson@gmail.com>

2018-04-18 20:04:06 +0200
    Improved clause regex. Removed req for leading 0 in decimal
    number.

        Benct Philip Jonsson <bpjonsson@gmail.com>

2018-04-18 18:26:43 +0200
    minor markup correction and README additions

        Benct Philip Jonsson <bpjonsson@gmail.com>

2018-04-18 17:53:49 +0200
    fix backslash escapes in character classes

        Benct Philip Jonsson <bpjonsson@gmail.com>

2018-04-18 17:39:36 +0200
    add README.md and increase version number

        Benct Philip Jonsson <bpjonsson@gmail.com>

2018-04-18 16:23:55 +0200
    added .gitignore

        Benct Philip Jonsson <bpjonsson@gmail.com>

2018-04-18 16:17:59 +0200
    Document and improve regex attribute name matching.

        Benct Philip Jonsson <bpjonsson@gmail.com>

    Rename files.

    General syntax documentation cleanup.

    Allow internal hyphens in unquoted strings.

2018-04-17 16:53:23 +0200
    correct minimum perl version in .pl file

        Benct Philip Jonsson <bpjonsson@gmail.com>

2018-04-17 14:47:23 +0200
    minor correction in syntax description

        Benct Philip Jonsson <bpjonsson@gmail.com>

2017-02-??
    first version

        Benct Philip Jonsson <bpjonsson@gmail.com>


pandoc-elements-selectors.pl
#!/usr/bin/env perl

# Proposed Pandoc::Elements extended selector expression syntax
#
# VERSION 201804171651
#
# You can test the syntax by running this script and typing selectors to STDIN.
# The selector will be compiled and the compiled subroutine or any error message
# will be printed to STDOUT.

# use utf8;      
use utf8::all;
use autodie 2.26;
use 5.010001;     
use strict;    
use warnings;  
use warnings  qw(FATAL utf8);    
# use open      qw(:std :utf8);    
# use charnames qw(:full :short);  

use Carp               qw[ carp croak confess cluck     ];

use Data::Printer deparse => 1, alias => 'ddp', output => 'stdout';

my $string_re = qr{
    (?:
        (?<quote> '   ) (?<string>  [^']*   (?: ''  [^']*   )* ) '
    |   (?<quote> "   ) (?<string>  [^"]*   (?: ""  [^"]*   )* ) "
    |   (?<quote> [{] ) (?<regex>   (?&REGEX)                  ) [}] (?<mod> \w* )
    |   (?<quote>     ) (?<string>  \b\w+\b )
    )
    (?(DEFINE)
        (?<REGEX>
            (?:
                [^\{\}\[\]\\]+                      # 'normal' characters
            |   \\.                                 # or backslash escapes
            |   \[ \]? [^\]]* (?: \\. [^\]]* )* \]  # or character classes
            |   [{] (?&REGEX) [}]                   # or nested balanced braces
            )*
        )
    )
}msx;

my $expression_re = qr{^$string_re$};

my $clause_re = qr{
    (?<not> \!? )
    (?:
        (?<sigil> \: )                                  (?<type>  document|block|inline|meta)
    |   (?<sigil> \& )  (?<method> (?&ID)  )     (?<cmp> (?&CMP) ) (?<num>   (?&NUM)    )
    |   (?<sigil> \& )  (?<method> (?&ID)  ) (?: (?<cmp> (?&OP)  ) (?<value> (?&STRING) ) )?
    |   (?<sigil> \% )  (?<key> (?&STRING) )     (?<cmp> (?&CMP) ) (?<num>   (?&NUM)    )
    |   (?<sigil> \% )  (?<key> (?&STRING) ) (?: (?<cmp> (?&OP)  ) (?<value> (?&STRING) ) )?
    |   (?<sigil> \# )                                             (?<value> (?&STRING) )
    |   (?<sigil> \. )                                             (?<value> (?&STRING) )
    |   (?<sigil>    )                                             (?<value> (?&STRING) )
    )
    \s* 
    (?(DEFINE)
        (?<ID> (?!\d) \w+ )
        (?<STRING> $string_re )
        (?<NUM> -?\b[0-9]+(?: \. [0-9]+\b )? )
        (?<CMP> [\!\=]?\= | [\<\>]\=? )
        (?<OP>  [\!\=]?\~ )
    )
}msx;

my $selector_re = qr{
    \s* (?<selector> (?&CLAUSE)+ ) \s* (?: \| \s* | \z ) | (?<invalid> \S+ )
    (?(DEFINE) (?<CLAUSE> $clause_re ) )
}msx;

my $compile_quoted = sub {
    my($match) = @_;
    my $quote  = $match->{quote};
    (my $string = $match->{string}) =~ s/[$quote]{2}/$quote/g;
    # ddp $string;
    return qr{^\Q$string\E$};
};

my %compile_expression = (
    q[']    => $compile_quoted,
    q["]    => $compile_quoted,
    q[]     => sub { qr{^\Q$_[0]{string}\E$} },
    q[{]    => sub {
        my($match) = @_;
        my $regex = $match->{regex};
        my $mod = $match->{mod};
        my $ret;
        $regex =~ s{\\(.)|([\$\@\/])}{\\$+}g;
        eval { $ret = qr/(?$mod:$regex)/; 1; } or do{
            my $e = $@;
            croak "Error compiling regex in selector:\n$e";
        };
        return $ret;
    },
);

my $compile_method_check = sub {
    state $method_for_sigil = +{
        q{#} => q/id/,
    };
    my ( $match ) = @_;
    my ( $m, $n, $c, $v) = map {; $_ // "" } @{$match}{qw[method not cmp value]};
    $c ||= '=~';
    $m ||= $method_for_sigil->{$match->{sigil}};
    if ( length $v ) {
        return qq{ ( $n( \$e->can('$m') and (\$e->$m $c $v) ) ) };
    }
    else {
        return qq{ ( $n\$e->can('$m') ) };
    }
};

my $compile_kv_check = sub {
    state $key_for_sigil = +{
        q{.} => q/^class$/,
    };
    my($match) = @_;
    $match->{key} //= $key_for_sigil->{$match->{sigil}};
    my($k, $v, $n, $c) = map {; $_ // "" } @{$match}{qw[key value not cmp]};
    $c ||= '=~';
    if ( length $v ) {
        $v = qq{ \$_ $c $v };
    }
    if ( length $v ) {
        return qq{
            (   $n(
                    \$e->can( 'keyvals' ) and do {
                        my \$kv   = \$e->keyvals;
                        my \$hits = 0;
                        for my \$key ( grep { /$k/ } \$kv->keys ) {
                            \$hits += ( grep { $v } \$kv->get_all( \$key ) );
                        }
                        !!\$hits;
                    }
                )
            )
        };
    }
    else {
        return qq{
            ( $n( \$e->can( 'keyvals' ) and !!(grep { /$k/ } \$e->keyvals->keys)))
        };
    }
};

my %compile_clause = (
    q{:} => sub {
        my ( $match ) = @_;
        my $t         = "is_$match->{type}";
        my $n         = $match->{not};
        return qq{ ( $n(\$e->$t) ) };
    },
    q{} => sub {
        my ( $match ) = @_;
        my ( $n, $v ) = @{$match}{qw[not value]};
        return qq{ ( $n(\$e->name =~ $v) ) };
    },
    q{&} => $compile_method_check,
    q{#} => $compile_method_check,
    q{.} => $compile_kv_check,
    q{%} => $compile_kv_check,
);

sub _compile_selector {
    my($string) = @_;
    my @selectors;
    while ( $string =~ /\G$selector_re/gc ) {
        my %match = %+;
        my @pos = @-;
        if ( length $match{invalid} ) {
            my $fill = substr $string, 0, $pos[2];
            $fill =~ s{ \X }{ }gx;
            chomp $string;
            croak "Invalid expression in selector marked by '^':\n\n$string\n$fill^\n";
        }
        push @selectors, $match{selector}
    }
    for my $selector ( @selectors ) {
        $selector = _compile_clauses($selector);
    }
    my $selectors_code = join qq{ )\nor( }, @selectors;
    my $code = qq{ \$sub = sub { my(\$e) = \@_; return( ( $selectors_code ) ); }; 1; };
    my $sub;
    local $@;
    eval($code) || do {
        my $e = $@;
        croak "Error compiling selector:\n$string\n\n$code\n\n$e\n";
    };
    return $sub;
}

sub _compile_clauses {
    my( $selector ) = @_;
    my @clauses;
    while( $selector =~ /\G$clause_re/gc ) {
        my $match = +{ %+ };
        for my $key ( qw[ key value ] ) {
            next unless defined $match->{$key};
            next unless $match->{$key} =~ /$expression_re/;
            $match->{$key} = $compile_expression{$+{quote}}->(+{ %+ });
        }
        if ( $match->{cmp} ) {
            $match->{cmp} =~ s/^\~/=~/;
            $match->{cmp} =~ s/^=(?![=~])/==/;
        }
        if ( $match->{num} ) {
            ## Ensure decimal interpretation!
            $match->{num} =~ s/^0+(?=\d)//;
        }
        $match->{value} &&= "/$match->{value}/";
        $match->{value} ||= $match->{num};
        push @clauses, $compile_clause{ $match->{sigil} }->($match);
    }
    return join 'and', @clauses;
}

 
use Try::Tiny;

while ( <> ) {
    chomp;
    ddp my $sub = try { _compile_selector($_) } catch { $_ };
}


pandoc-elements-selectors-parser.pl
#!/usr/bin/env perl

# Proposed Pandoc::Elements extended selector expression syntax
#
# VERSION 201805021600
#
# The latest version of this script can be found at 
# <https://github.com/bpj/pandoc-elements-selectors>
#
# You can test the syntax by running this script and typing selectors to STDIN.
# The selector will be compiled and the compiled subroutine or any error message
# will be printed to STDOUT.

# use utf8;      
use utf8::all;
use autodie 2.26;
use 5.010001;     
use strict;    
use warnings;  
use warnings  qw(FATAL utf8);    
# use open      qw(:std :utf8);    
# use charnames qw(:full :short);  

use Carp               qw[ carp croak confess cluck     ];

use Data::Printer deparse => 1, alias => 'ddp', output => 'stdout';

my $string_re = qr{
    (?:
        (?<quote> '   ) (?<string>  [^']*   (?: ''  [^']*   )* ) '
    |   (?<quote> "   ) (?<string>  [^"]*   (?: ""  [^"]*   )* ) "
    |   (?<quote> [{] ) (?<regex>   (?&REGEX)                  ) [}] (?<mod> \w* )
    |   (?<quote>     ) (?<string>  \b(?=\w)[-\w]*\w\b         )
    )
    (?(DEFINE)
        (?<REGEX>
            (?:
                [^\{\}\[\]\\]+                          # 'normal' characters
            |   \\.                                     # or backslash escapes
            |   \[ \]? [^\]\\]* (?: \\. [^\]\\]* )* \]  # or character classes
            |   [{] (?&REGEX) [}]                       # or nested balanced braces
            )*
        )
    )
}msx;

my $word_re = qr{^$string_re$};

my $expression_re = qr{
    (?<not> \!? )
    (?:
        (?<sigil> \: ) (?<type> document|block|inline|meta )
    |   (?<sigil> \& )                  (?<method> (?&ID)  ) (?: (?<cmp> (?&CMP) ) (?<num>   (?&NUM)    )
                                                             |   (?<cmp> (?&OP)  ) (?<value> (?&STRING) )
                                                             )?
    |   (?<sigil> \% ) (?<op> (?&OP)? ) (?<key> (?&STRING) ) (?: (?<cmp> (?&CMP) ) (?<num>   (?&NUM)    )
                                                             |   (?<cmp> (?&OP)  ) (?<value> (?&STRING) )
                                                             )?
    |   (?<sigil> [.#]? )                                                          (?<value> (?&STRING) )
    )
    \s* 
    (?(DEFINE)
        (?<ID> (?!\d) \w+ )
        (?<NUM> -? \b (?: [0-9]* \. [0-9]+ | [0-9]+ ) \b )
        (?<CMP> [\!\=]?\= | [\<\>]\=? )
        (?<OP>  [\!\=]?\~ )
        (?<STRING> $string_re )
    )
}msx;

my $selector_re = qr{
    \s* (?<selector> (?&EXPRESSION)+ ) \s* (?: \| \s* | \z ) | (?<invalid> \S+ )
    (?(DEFINE) (?<EXPRESSION> $expression_re ) )
}msx;

my $compile_quoted = sub {
    my($match) = @_;
    my $quote  = $match->{quote};
    (my $string = $match->{string}) =~ s/[$quote]{2}/$quote/g;
    return qr{^\Q$string\E$};
};

my %compile_word = (
    q[']    => $compile_quoted,
    q["]    => $compile_quoted,
    q[]     => sub { qr{^\Q$_[0]{string}\E$} },
    q[{]    => sub {
        my($match) = @_;
        my $regex = $match->{regex};
        my $mod = $match->{mod};
        my $ret;
        $regex =~ s{\\(.)|([\$\@\/])}{\\$+}g;
        eval { $ret = qr/(?$mod:$regex)/; 1; } or do{
            my $e = $@;
            croak "Error compiling regex in selector:\n$e";
        };
        return $ret;
    },
);

my $compile_method_check = sub {
    state $method_for_sigil = +{
        q{#} => q/id/,
    };
    my ( $match ) = @_;
    my ( $m, $n, $c, $v) = map {; $_ // "" } @{$match}{qw[method not cmp value]};
    $c ||= '=~';
    $m ||= $method_for_sigil->{$match->{sigil}};
    if ( length $v ) {
        return qq{ ( $n( \$e->can('$m') and (\$e->$m $c $v) ) ) };
    }
    else {
        return qq{ ( $n\$e->can('$m') ) };
    }
};

my $compile_kv_check = sub {
    state $key_for_sigil = +{
        q{.} => q{/^class$/},
    };
    my($match) = @_;
    $match->{key} //= $key_for_sigil->{$match->{sigil}};
    my($k, $v, $n, $c, $o) = map {; $_ // "" } @{$match}{qw[key value not cmp op]};
    for my $x ( [ \$k, \$o ], [ \$v, \$c ] ) {
        ${$x->[1]} ||= '=~';
        if ( length ${$x->[0]} ) {
            ${$x->[0]} = qq{ \$_ ${$x->[1]} ${$x->[0]} };
        }
    }
    if ( length $v ) {
        return qq{
            (   $n(
                    \$e->can( 'keyvals' ) and do {
                        my \$kv   = \$e->keyvals;
                        my \$hits = 0;
                        for my \$key ( grep { $k } \$kv->keys ) {
                            \$hits += ( grep { $v } \$kv->get_all( \$key ) );
                        }
                        !!\$hits;
                    }
                )
            )
        };
    }
    else {
        return qq{
            ( $n( \$e->can( 'keyvals' ) and !!(grep { $k } \$e->keyvals->keys)))
        };
    }
};

my %compile_expression = (
    q{:} => sub {
        my ( $match ) = @_;
        my $t         = "is_$match->{type}";
        my $n         = $match->{not};
        return qq{ ( $n(\$e->$t) ) };
    },
    q{} => sub {
        my ( $match ) = @_;
        my ( $n, $v ) = @{$match}{qw[not value]};
        return qq{ ( $n(\$e->name =~ $v) ) };
    },
    q{&} => $compile_method_check,
    q{#} => $compile_method_check,
    q{.} => $compile_kv_check,
    q{%} => $compile_kv_check,
);

sub _compile_selector {
    my($string) = @_;
    my @selectors;
    while ( $string =~ /\G$selector_re/gc ) {
        my %match = %+;
        my @pos = @-;
        if ( length $match{invalid} ) {
            my $fill = substr $string, 0, $pos[2];
            $fill =~ s{ \X }{ }gx;
            chomp $string;
            croak "Invalid expression in selector marked by '^':\n\n$string\n$fill^\n";
        }
        push @selectors, $match{selector}
    }
    for my $selector ( @selectors ) {
        $selector = _compile_expressions($selector);
    }
    my $selectors_code = join qq{ )\nor( }, @selectors;
    my $code = qq{ \$sub = sub { 
        no warnings qw[ uninitialized numeric ];
        my(\$e) = \@_; return( ( $selectors_code ) ); };
        1; 
    };
    my $sub;
    local $@;
    eval($code) || do {
        my $e = $@;
        croak "Error compiling selector:\n$string\n\n$code\n\n$e\n";
    };
    return $sub;
}

sub _compile_expressions {
    my( $selector ) = @_;
    my @expressions;
    while( $selector =~ /\G$expression_re/gc ) {
        my $match = +{ %+ };
        for my $key ( qw[ key value ] ) {
            next unless defined $match->{$key};
            next unless $match->{$key} =~ /$word_re/;
            $match->{$key} = $compile_word{$+{quote}}->(+{ %+ });
            $match->{$key} = "/$match->{$key}/"
        }
        for my $cmp ( qw[ cmp op ] ) {
            $match->{$cmp} // next;
            $match->{$cmp} ||= '=~';
            $match->{$cmp} =~ s/^\~/=~/;
            $match->{$cmp} =~ s/^=(?![=~])/==/;
        }
        if ( $match->{num} ) {
            ## Ensure decimal interpretation!
            $match->{num} =~ s/^0+(?=\d)//;
        }
        $match->{value} ||= $match->{num};
        push @expressions, $compile_expression{ $match->{sigil} }->($match);
    }
    return join 'and', @expressions;
}

 
use Try::Tiny;

while ( <> ) {
    chomp;
    ddp my $sub = try { _compile_selector($_) } catch { $_ };
}
